python/ml4ir/config/features.py

```diff
@@ -1,120 +1,237 @@
+
+        try:
+            self.query_key = features_dict.get(FeatureConfigKey.QUERY_KEY)
+            self.all_features.append(self.query_key)
+        except KeyError:
+            self.query_key = None
+            if logger:
+                logger.warning("'query_key' key not found in the feature_config specified")
+
+        try:
+            self.rank = features_dict.get(FeatureConfigKey.RANK)
+            self.all_features.append(self.rank)
+        except KeyError:
+            self.rank = None
+            if logger:
+                logger.warning("'rank' key not found in the feature_config specified")
+
+        try:
+            self.label = features_dict.get(FeatureConfigKey.LABEL)
+            self.all_features.append(self.label)
+        except KeyError:
+            raise KeyError("'label' key not found in the feature_config specified")
+
+        try:
+            self.features: List = features_dict.get(FeatureConfigKey.FEATURES)
+            self.all_features.extend(self.features)
+        except KeyError:
+            raise KeyError("'features' key not found in the feature_config specified")
+
+        # Features that can be used for training the model
+        self.ranking_features = list()
+        # Features that provide additional information about the query+records
+        self.metadata_features = list()
+        # Features that contain information at the query level common to all records
+        self.context_features = list()
+        # Features that contain information at the record level
+        self.sequence_features = list()
+        # Features to log at inference time
+        self.features_to_log = list()
+        for feature_info in self.all_features:
+            if feature_info.get("trainable", True):
+                self.ranking_features.append(feature_info)
+            else:
+                self.metadata_features.append(feature_info)
+
+            if feature_info.get("tfrecord_type") == TFRecordTypeKey.CONTEXT:
+                self.context_features.append(feature_info)
+            elif feature_info.get("tfrecord_type") == TFRecordTypeKey.SEQUENCE:
+                self.sequence_features.append(feature_info)
+            else:
+                raise TypeError(
+                    "tfrecord_type should be either context or sequence "
+                    "but is {} for {}".format(
+                        feature_info.get("tfrecord_type"), feature_info.get("name")
+                    )
+                )
+
+            if feature_info.get("log_at_inference", False):
+                self.features_to_log.append(feature_info)
+
+        if len(self.ranking_features) == 0:
             raise Exception("No trainable features specified in the feature config")
-        if not self.label:
-            raise Exception("No label field defined in the feature config")
 
-    def get_feature_names(self, trainable=None, include_label=False, dtype=None) -> List[str]:
+        if logger:
```

Comment:
probably separate this whole if block into a logInitialization() method, to avoid clutter in the constructor